plutus-core/plutus-ir/src/PlutusIR/Transform/Inline/Heuristics.hs

```diff
@@ -0,0 +1,253 @@
+    ( HasUnique name TermUnique
+    , HasUnique tyname TypeUnique
+    , Eq name
+    , PLC.ToBuiltinMeaning uni fun
+    )
+
+data InlineInfo name fun a = InlineInfo
+    { _iiStrictnessMap :: Deps.StrictnessMap
+    -- ^ Is it strict? Only needed for PIR, not UPLC
+    , _iiUsages        :: Usages.Usages
+    -- ^ how many times is it used?
+    , _iiHints         :: InlineHints name a
+    -- ^ have we explicitly been told to inline.
+    , _iiBuiltinVer    :: PLC.BuiltinVersion fun
+    -- ^ the builtin version.
+    }
+makeLenses ''InlineInfo
+
+-- Using a concrete monad makes a very large difference to the performance of this module
+-- (determined from profiling)
+-- | The monad the inliner runs in.
+type InlineM tyname name uni fun a =
+    ReaderT (InlineInfo name fun a) (StateT (Subst tyname name uni fun a) Quote)
+
+-- | Check if term is pure. See Note [Inlining and purity]
+checkPurity
+    :: forall tyname name uni fun a. InliningConstraints tyname name uni fun
+    =>Term tyname name uni fun a -> InlineM tyname name uni fun a Bool
+checkPurity t = do
+    strctMap <- view iiStrictnessMap
+    builtinVer <- view iiBuiltinVer
+    let strictnessFun = \n' -> Map.findWithDefault NonStrict (n' ^. theUnique) strctMap
+    pure $ isPure builtinVer strictnessFun t
+
+nameUsedAtMostOnce :: forall tyname name uni fun a. InliningConstraints tyname name uni fun
+    => name
+    -> InlineM tyname name uni fun a Bool
+nameUsedAtMostOnce n = do
+    usgs <- view iiUsages
+    -- 'inlining' terms used 0 times is a cheap way to remove dead code while we're here
+    pure $ Usages.getUsageCount n usgs <= 1
+
+effectSafe :: forall tyname name uni fun a. InliningConstraints tyname name uni fun
+    => Term tyname name uni fun a
+    -> Strictness
+    -> name
+    -> Bool -- ^ is it pure?
+    -> InlineM tyname name uni fun a Bool
+effectSafe body s n purity = do
+    -- This can in the worst case traverse a lot of the term, which could lead to us
+    -- doing ~quadratic work as we process the program. However in practice most term
+    -- types will make it give up, so it's not too bad.
+    let immediatelyEvaluated = case firstEffectfulTerm body of
+            Just (Var _ n') -> n == n'
+            _               -> False
+    pure $ case s of
+        Strict    -> purity || immediatelyEvaluated
+        NonStrict -> True
+
+-- | Should we inline? Should only inline things that won't duplicate work or code.
+-- See Note [Inlining approach and 'Secrets of the GHC Inliner']
+acceptable ::  Term tyname name uni fun a -> InlineM tyname name uni fun a Bool
+acceptable t =
+    -- See Note [Inlining criteria]
+    pure $ costIsAcceptable t && sizeIsAcceptable t
+
+{- |
+Try to identify the first sub term which will be evaluated in the given term and
+which could have an effect. 'Nothing' indicates that we don't know, this function
+is conservative.
+-}
+firstEffectfulTerm :: Term tyname name uni fun a -> Maybe (Term tyname name uni fun a)
```

Comment:
This is a generic utility.

Reply:
Moved to `Purity` module. It's kinda related...